modal-go/auth_token_manager.go

```diff
@@ -0,0 +1,203 @@
+	token := m.GetCurrentToken()
+
+	// Return an existing, non-expired token.
+	if token != "" && !m.IsExpired() {
+		return token, nil
+	}
+
+	// Should almost never happen.
+	return m.FetchToken(ctx)
+}
+
+// backgroundRefresh runs in a goroutine and refreshes tokens REFRESH_WINDOW seconds before they expire.
+func (m *AuthTokenManager) backgroundRefresh(ctx context.Context) {
+	// If we have no token, fetch one.
+	if m.GetCurrentToken() == "" {
+		if _, err := m.FetchToken(ctx); err != nil {
+			fmt.Printf("Failed to fetch auth token: %v\n", err)
+		}
+	}
+
+	for {
+		expiry := m.GetExpiry()
+		now := time.Now().Unix()
+		refreshTime := expiry - REFRESH_WINDOW
+
+		var delay time.Duration
+		if refreshTime > now {
+			// Token does not need refreshing yet. Set a delay to wait until the refresh time.
+			delay = time.Duration(refreshTime-now) * time.Second
+		} else {
+			// Token needs refresh now or is expired, refresh immediately
+			// This should almost never happen.
+			delay = 0
+		}
+
+		if delay > 0 {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(delay):
+			}
+		}
+
+		// Refresh the token
+		if _, err := m.FetchToken(ctx); err != nil {
+			fmt.Printf("Failed to refresh auth token: %v\n", err)
+		}
+	}
+}
+
+// requestToken fetches a new token from the server.
+func (m *AuthTokenManager) requestToken(ctx context.Context) (string, error) {
+	resp, err := m.client.AuthTokenGet(ctx, &pb.AuthTokenGetRequest{})
+	if err != nil {
+		return "", fmt.Errorf("failed to get auth token: %w", err)
+	}
+
+	token := resp.GetToken()
+	if token == "" {
+		return "", fmt.Errorf("internal error: did not receive auth token from server, please contact Modal support")
+	}
+
+	m.token.Store(token)
+
+	if exp := m.DecodeJWT(token); exp > 0 {
+		m.expiry.Store(exp)
+	} else {
+		fmt.Printf("Failed to decode x-modal-auth-token exp field\n")
+		// We'll use the token, and set the expiry to 20 min from now.
+		m.expiry.Store(time.Now().Unix() + DEFAULT_EXPIRY_OFFSET)
+	}
+
+	return token, nil
+}
+
+// Extracts the exp claim from a JWT token.
+func (m *AuthTokenManager) decodeJWT(token string) int64 {
+	parts := strings.Split(token, ".")
+	if len(parts) != 3 {
+		return 0
+	}
+
+	payload := parts[1]
+	for len(payload)%4 != 0 {
+		payload += "="
+	}
+
+	decoded, err := base64.URLEncoding.DecodeString(payload)
+	if err != nil {
+		return 0
+	}
+
+	var claims map[string]interface{}
+	if err := json.Unmarshal(decoded, &claims); err != nil {
+		return 0
+	}
+
+	if exp, ok := claims["exp"].(float64); ok {
+		return int64(exp)
+	}
+
+	return 0
+}
+
+// Sets the token and expiry.
+func (m *AuthTokenManager) SetToken(token string, expiry int64) {
+	m.token.Store(token)
+	m.expiry.Store(expiry)
+}
+
+// Returns the current token expiry.
+func (m *AuthTokenManager) GetExpiry() int64 {
+	return m.expiry.Load()
+}
+
+// Extracts the expiry timestamp from a JWT.
+func (m *AuthTokenManager) DecodeJWT(token string) int64 {
```

Comment:
Do we need these public wrapper functions? Might be more clear to call the private function directly?